Better clicking of the ng-select input target

oliverguenther opened the pull request:
This seemingly got broken in https://github.com/opf/openproject/commit/56abad9b2257eaf7667a5c5ec006196af85611d9#diff-50004f6a02383d357c8e4fdc4d91442c89d0160410a7bbf2f928e4e0dc99baf2

oliverguenther committed:
Better clicking of the target

klaustopher commented:
now filter specs seem to fail

oliverguenther committed:
Use target itself

myabc commented:
@oliverguenther @klaustopher FYI: Capybara Accessible Selectors [provides a `combo_box` selector](https://github.com/citizensadvice/capybara_accessible_selectors/tree/main?tab=readme-ov-file#combo_box), based on the ARIA `combobox` ARIA role. Finding comboboxes on the pages currently works, but the selector needs to be patched to assert and select options with `ng-select`. I hope to get a patch in upstream. / @bsatarnejad

myabc commented on spec/support/components/autocompleter/ng_select_autocomplete_helpers.rb:
```diff
@@ -4,8 +4,7 @@ module Components::Autocompleter
     def search_autocomplete(element, query:, results_selector: nil, wait_dropdown_open: true, wait_for_fetched_options: true)
       SeleniumHubWaiter.wait unless using_cuprite?
-      # Open the element
-      page.execute_script("arguments[0].click();", element.native)
```
I'm not sure why this change was made in the first place?

cbliard replied:
hard to track because the changes were stashed, but it's because of these commits: https://github.com/opf/openproject/pull/19405/commits/0fb4adc7216244c0b6b886df72edbeb051911157 and then https://github.com/opf/openproject/pull/19405/commits/504af2e151ea3e054c2efad90d405b5adbc6d258

> ensure reliable click on autocomplete by using JS click to avoid overlapping elements

oliverguenther committed:
Use trigger click for cuprite

cbliard commented:
Which tests is it supposed to fix?
I'm asking because I wanted to have a look at `./spec/features/work_packages/tabs/relations_children_spec.rb:129` which is failing consistently locally for me, and it does not seem to be fixed by this PR either. So if it's not the tests this PR is supposed to fix then I can try fixing it?

oliverguenther commented:
@cbliard I was trying to fix the relation specs, which works with the first commit. However it breaks other specs

cbliard commented:
ok I get it now.
without the explicit names of the specs, it's hard to follow.

cbliard committed:
Restore direct click and work around issue with other specs

The `./spec/features/work_packages/tabs/relations_children_spec.rb`
works better when using `target.click` instead of
`target.trigger('click')`.

The `./spec/features/work_packages/progress_modal_spec.rb` was scrolling
to the middle of the screen when clicking on the work field to display
the progress popover for some reason.

Then when updating the status field on the work package creation page,
it was made active by clicking it, which displayed the dropdown, and
then it was clicked again for the autocompletion, which scrolled at the
top of the page for some reason. Problem is: the dropdown is fixed and
scrolling up make it overlap the status field, and when calling
`target.click`, Capybara refuses saying that something overlaps it.

The adopted solution is to scroll to the top of the page before updating
the status. Not satisfying, but works.

cbliard committed:
Fix visibility of `Pages::FullWorkPackageCreate#set_status`

specs:
- spec/features/work_packages/progress_modal_spec.rb:281
- spec/features/work_packages/progress_modal_spec.rb:309
run: https://github.com/opf/openproject/actions/runs/16223078904/job/45808262760

cbliard committed:
Fix retrieval of email in users autocompleter

spec: spec/features/work_packages/edit_work_package_spec.rb:274
run: https://github.com/opf/openproject/actions/runs/16223078904/job/45808262760

Modifying `el.all(...).first` to `el.first(...)` was an error because
the first one returns `nil` if no element is found, while the second one
raises an error.

Adapted the spec to use `el.first(...)` but only after checking that the
element is present.

Also set `wait: 0` to avoid waiting for nothing when the element is not
there. We already have the autocompleter elements so waiting is not
needed here.

cbliard committed:
Try to wait for the ng-select to have loaded options before reading them

spec: spec/features/projects/lists/filters_spec.rb:509
run: https://github.com/opf/openproject/actions/runs/16223078904/job/45808262760

Trying to avoid having a `Capybara::Cuprite::ObsoleteNode` error.

cbliard commented:
`modules/meeting/spec/features/meetings_index_spec.rb:202` fails if it is run on a Saturday, because Sunday is the beginning of next week, so it is grouped in the "Next week and later" section instead of the "Tomorrow" section, which makes the test fail.

That's debatable, but I think "Tomorrow" should have precedence over "Next week".
Like instead of 
```ruby
    next_week = Time
      .current
      .next_occurring(OpenProject::Internationalization::Date.beginning_of_week)
      .beginning_of_day
```
have something like
```ruby
    next_week = Time
      .current
      .next_occurring(OpenProject::Internationalization::Date.beginning_of_week)
      .beginning_of_day
    after_tomorrow = Time.current.next_day(2).beginning_of_day
    next_week_after_tomorrow = [next_week, after_tomorrow].max
```

myabc commented:
Please also give #19545 a look
